### ISSUE TITLE
4.042 improperly encoding blobs when sql_type is SQL_UNKNOWN_TYPE

### ISSUE THREAD

**Issue by cthulhuology** [perl5-dbi/DBD-mysql#117]
The utf8 encoding changes have resulted in a regression that has unexpected side effects.
Consider the following script:

`use DBI;

#
# create table foo (foo longblob);
#

my $dbh = DBI->connect('DBI:mysql:database=test','desktop','',{ mysql_enable_utf8 => 1 })
    or die $DBI::errstr;

my $sth = $dbh->prepare('INSERT INTO foo (foo) values (?)');
$sth->execute("I18N Web Testing æøå");

my $rth = $dbh->prepare('SELECT foo FROM foo');
$rth->execute();
while (@row = $rth->fetchrow_array) {
    print $row[0],"\n";
}`

Now the longblob mysql type should be considered a SQL_BLOB, but because the exec call doesn't call bind_param with the attrib set to SQL_BLOB, the default value of 0 is used SQL_UNKNOWN_TYPE and passed to bind_param.  If you sv_dump at 939 in dbdimp.c you'll see:

`SV = PV(0x162d950) at 0x15f8630
  REFCNT = 1
  FLAGS = (POK,IsCOW,READONLY,PROTECT,pPOK)
  PV = 0x15dc200 "I18N Web Testing \303\246\303\270\303\245"\0
  CUR = 23
  LEN = 25
  COW_REFCNT = 0
`
And sql_type is the default value of 0.

the output of the scrip will print the value stored in the database:
`
I18N Web Testing Ã¦Ã¸Ã¥`
`
because it is double encoding the characters, and not preserving the blob value because the
sql_type_is_binary check returns false because the SQL_UNKNOWN_TYPE value of sql_type.

One example of widely used code that exhibit this behavior is Apache::Session (Apache::Session::MySQL), which calls bind_param explicitly but never passes the SQL_BLOB at and of the call sites.  Since Apache::Session is using a Storable (which is binary data stored in a blob) the additional UTF8 encoding can result in data corruption which can cause perl to crash with an OOM error when materialize is invoked.  This is a potential security threat.

**Comment by pali** [comment]
The whole `mysql_enable_utf8` was just experimental and totally broken prior to DBD::mysql 4.042. Experimental status was also written in documentation, see: https://metacpan.org/pod/release/CAPTTOFU/DBD-mysql-4.033/lib/DBD/mysql.pm#mysql_enable_utf8

I'm not sure what we can done here. Basically applications which trying to use experimental broken option provided by DBD::mysql < 4.042 and just broken.

Due to some fundamentals of perl scalars and also MySQL protocol itself, DBD::mysql does not type of bind parameter, so driver can use logic which you already described. When mysql_enable_utf8 is 0, then nothing is encoded/decoded and so treated as binary sequence of octets.

**Comment by Grinnz** [comment]
In other words: DBD::mysql does not know that a parameter is being used for insert into a blob column. Since you set mysql_enable_utf8 it will encode it, unless bound with the SQL_BLOB or similar type. On retrieval it knows that the column is a binary type, so it does not decode it.

**Comment by ap** [comment]
> The whole `mysql_enable_utf8` was just experimental and totally broken prior to DBD::mysql 4.042.

I used it for 10 years without problems. I’m sure I *would* have had problems if I had tried things where the incongruity of its design got exposed, but I never needed to. So my code had no bugs – even though DBD::mysql was designed incorrectly.

Worse: <http://grep.cpan.me/?q=mysql_enable_utf8+file%3A.pm> There’s plenty of infrastructure-level code just on CPAN alone which sets this flag *on behalf of* the user.

In the face of these realities, the fact that the documentation declared the feature experimental back when it was introduced, 11 years ago, is simply irrelevant today.

But now `mysql_enable_utf8` is fixed – I‘m not going argue about that. But users’ code that wasn’t broken before now is. And this was simply unnecessary.

**Comment by pali** [comment]
So... what do you suggest? As wrote I have no idea what can be done here. Probably the best solution would be to fix broken code.

And about `mysql_enable_utf8` usage on cpan, you do not know how many of those modules are broken and how many expect that `mysql_enable_utf8` is not buggy? If module is multi-dbd and set `mysql_enable_utf8`, `pg_enable_utf8` and also `sqlite_unicode` then it expects that `mysql_enable_utf8` is working (which means needs DBD::mysql >= 4.042).

**Comment by ap** [comment]
Of course many of these CPAN modules will produce broken results when used in any somewhat unusual configuration where the broken design actually matters. But a lot of people don’t try that, and for them it ends up not mattering that the design is not sound.

I’m not sure what I’d suggest specifically. The general idea is always the same though, don’t *change* how `mysql_enable_utf8` works – for now at least –, but *add* the more correct encoding approach, and then give users a way to ask for it.

How exactly you do that depends on the code and what you need to achieve long term. New flag e.g. `mysql_enable_encoding`? New flag value e.g. `mysql_enable_utf8 => 2`? Some other approach maybe? Also, do you warn when `mysql_enable_utf8` is used? Maybe detect problematic uses and only warn then? Overall I would weigh that choice based on how difficult it is to support both the old and new thing long-term. The easier it is to keep the old stuff around, the less old users should be pushed away from it.

(Of course when it comes to the docs, the old approach should be de-emphasised, and the new way should be shown as the obvious/standard way of doing things, to steer new users in that direction. But the old way must stay documented – because there is still code that uses it, and people who have to maintain that code need to be able to figure out what it does. Also, to move users toward the new way it is also important to make sure that the new way doesn’t make too many things harder to do than the old way (esp. commonly needed things).)

That is the general idea.

**Comment by pali** [comment]
Problem with old `mysql_enable_utf8` implementation is that it behave "pseudo-randomly". For frozen combination of perl at specific version, all cpan modules at specific version and user application it acts deterministic. But once either perl or any cpan module is updated old `mysql_enable_utf8` (prior to 4.042) can change result of DBD::mysql. So result is basically unpredictable.

If users (and authors of other modules) are not going to fix broken code and need current behavior they should stay at specific version of perl and all cpan modules. Otherwise it is not guaranteed that also old DBD::mysql version (prior to 4.042) provide same results. It can lead to results: after updating cpan module XYZ to new version MySQL server in user application started returning wrong result, where XYZ can be module which has nothing to do with MySQL.

Supporting such pseudo-random behavior is not easy and funny.

And I really suggest to conserve such fragile code and disallow updating code around (which can fully break it).

What is probably more important: how to write `mysql_enable_utf8` code which tries to be compatible with both old buggy DBD::mysql versions and new.

I created this pull request which adds "hacks" into documentation: https://github.com/perl5-dbi/DBD-mysql/pull/119

Please comment, test or provide another suggestion for it. To improve situation.

**Comment by ap** [comment]
> But once either perl or any cpan module is updated old `mysql_enable_utf8` (prior to 4.042) can change result of DBD::mysql.

*Any* CPAN module? How?

**Comment by Grinnz** [comment]
If a CPAN module provides a string with a different internal representation, even if it's the same string, previous versions of DBD::mysql will treat it differently. We had this issue with output from Spreadsheet::ParseExcel that returned (perfectly normal) un-upgraded strings with byte values between 128 and 255, probably due to some dependency, causing errors when used as params to DBD::mysql with mysql_enable_utf8 set, as the internal representation it used was not UTF-8 encoded. Any module that provides you with a string could change what representation it gives you and break DBD::mysql while the string value is unchanged.

**Comment by pali** [comment]
@Grinnz is right. And moreover un-upgraded strings are created by default from string constants if you do not `use utf8` in your file from which those strings come from. Also some older perl versions generate un-upgraded string even if `\N{U+XX}` construction is used. And any perl function can upgrade or downgrade string as needed.

So please look at pull request https://github.com/perl5-dbi/DBD-mysql/pull/119 if it helps you or not...

**Comment by ap** [comment]
@Grinnz:

> If a CPAN module provides a string with a different internal representation, even if it's the same string, previous versions of DBD::mysql will treat it differently.

Yeah, but that isn’t DBD::mysql changing its behaviour. That’s some other module changing *its* behaviour. If the user of both modules compensates for the change in the other module, DBD::mysql will get the same inputs as before, and will behave the same way as before. The behaviour of DBD::mysql is not changing, just because of an upgrade to some other module.

It’s perfectly *predictable* what it does, even if it’s not correct. “When the inputs change, the behaviour changes; when the inputs are the same, the behaviour is the same” is the *opposite* of random.

> A more complete discussion of the issue is in https://rt.cpan.org/Ticket/Display.html?id=87428

Funny, there’s [an almost verbatim copy of my suggestion](https://rt.cpan.org/Public/Bug/Display.html?id=87428#txn-1347225) already in there. As far as I can tell, doing it that would have worked just fine… with no breakage. It could have been done that way and that would have given you what you wanted, without causing anyone else any pain. So why choose to cause pain when you don’t even gain anything from it?

@pali: So…

> Supporting such pseudo-random behavior is not easy and funny.

… if this meant what @Grinnz was talking about, then I disagree that that behaviour is random (even just pseudo), I disagree that it’s hard to support solely on that basis, and I disagree about what’s not funny here (namely, changing it).

> So please look at pull request #119 if it helps you or not...

That is indeed useful information in order to cope with the problems of the old interface. But I don’t see anything in there which would argue against adding the correct interface under a different flag – rather than changing `mysql_enable_utf8` and thereby breaking already-working code.

Has there been any technical argument for not using a different flag? So far, all I have seen are arguments about the correctness of not using the UTF8 flag – which I do not disagree with at all.

**Comment by pali** [comment]
> That’s some other module changing its behaviour.

I cannot agree here with you, because calling utf8::upgrade() or composing string via `\x{...}\x{...}` sequences, or via `\N{...}` is not behavior change. It is same behavior with exactly same function API.

What is changing here is low level representation, and you can compare it compiling program with different optimization flags, e.g. -O0 or -O2 in gcc. In both cases you get same program, just compiled with different assembler instructions.

And in perl usage of utf8::upgrade() or `\N{...}` is similar behavior. Perl just translate source file to different opcodes, but meaning of program is same...

Note that perl operators like `.` (dot, concatenation) or perl functions like substr calls utf8::upgrade() in some cases. And there is no guarantee that due to optimization in future it would not be changed (under which conditions is utf8::upgrade() called).

I wrote it is pseudo-random (not fully-random), which means it is deterministic. So I could agree with you that behavior can be predicted.

Originally I talk with DBD::mysql maintainers (half a year ago?) and wrote first proposed solution that introduction of new attribute for utf8 with hacking code around with supporting both of them can be possible. But IIRC we dismiss it due to introduction of big mess in code and reason that current behaviour does not make sense in perl.

Due to above fact and also that calling utf8::upgrade() is legal and already done by perl itself at lot of places and also that in documentation was explicitely written "experimental and may change" state, I introduced patches which fixes `mysql_enable_utf8`, rather then having two different UTF-8 supports... Note that I explicitely stated in pull request https://github.com/perl5-dbi/DBD-mysql/pull/67 that it would change behavior for more then month in review state and about 3 months there was on cpan engineering version. So I though it would be OK as nobody was against.

**Comment by sartak** [comment]
> Originally I talk with DBD::mysql maintainers (half a year ago?) and wrote first proposed solution that 
> introduction of new attribute for utf8 with hacking code around with supporting both of them can be 
> possible. But IIRC we dismiss it due to introduction of big mess in code and reason that current 
> behaviour does not make sense in perl.

Please reconsider this stance. In as much as the new behavior is "correct", the old behavior was widely deployed and has a tremendous amount of inertia. Migrating to the new behavior will require coordination and substantial effort. We (Request Tracker) are willing to undergo that effort, but we do need to prepare for it. Given that we are in the middle of a stable series and DBD-mysql 4.042 is getting traction, we are unable to do such a migration right now. Our only recourse right now is to refuse to install under 4.042, forcing users to install 4.041 by hand.

**Comment by pali** [comment]
> old behavior was widely deployed

What we learnt from this is that developers do not read documentation nor care about states like "This option is experimental and may change in future versions.".

Now I'm thinking how could be experimental features which are not expected to work correctly or which can be removed/changed at any time introduced into modules...

**Comment by ap** [comment]
> What we learnt from this is that developers do not read documentation nor care about states like "This option is experimental and may change in future versions.".

How seriously can you expect them to take that after the feature has been “experimental” without changes in behaviour for 10 years?

**Comment by sartak** [comment]
> How seriously can you expect them to take that after the feature has been “experimental” without changes in behaviour for 10 years?

Furthermore see this commit and especially its message: 026fa02d80afc7f3473e67b9f63fe1b74575da3e

**Comment by mbeijen** [comment]
Yeah, I removed the 'this is experimental' message last year because I knew lots of software depended at that point on this switch because it was the only way to get some form of sane behaviour from DBD::mysql - which had lots of other issues, but also mostly worked for most people.

I'm really not so happy with the situation that emerged. I though we fixed lots of smaller issues but I also thought we had enough unit test coverage to not break existing software using DBD::mysql for people. We explicitly sollicited feedback after posting development releases to CPAN by making blog posts on b.p.o, sending out mails to mailing lists, and of course we only got actual feedback (like this issue) after we released the stable.

@CaptTofu please advise!

**Comment by ap** [comment]
It doesn’t seem too late to add a `mysql_enable_unicode` with the new behaviour and put `mysql_enable_utf8` back the way it was. It’s been a couple months, but a couple months is not long: the cascade of packagers and package upgrades is much longer than that, so a lot of users are not yet affected.

**Comment by pali** [comment]
> How seriously can you expect them to take that after the feature has been “experimental” without changes in behaviour for 10 years?

Because there were open bugs that it does not work for 10 years?

**Comment by ap** [comment]
What software doesn’t have “some open bugs”? I do not believe you hold even yourself to that standard. If that was the bar for what software is safe to use… what would be left?

**Comment by nightlight1** [comment]
Hi, I'm an end user that would like to share with you the impact of this change.
I work at a company that has 100GB of files stored in MySQL databases of 5+ years old.
After upgrading this module all all user files that are inserted into tables with LONGBLOB columns are corrupt when they are retrieved.
The manual says that BLOB files are not encoded but it does (see testing code below).
Now we are challenged to find out which records are corrupt, and which ones are not.
Unsetting the flag is not an option as all our text data is utf8 mixed with LONGBLOBs.
The proposed solutions asks for a review of every written SQL statement and start specifying for each column its DataType.
This is not user friendly as it is more time consuming to write code.
This change breaks with Perl's paradigm that backwards compatibility is safeguarded by the habbit to test all changes against all exiting CPAN modules before releasing it.
The arguments from this change are understandable from a technical point of view, but we prefer to maintain usability by keeping the module backwards compatible.
Thank you for your understanding.

Manual text: 
> Input bind parameters of binary types (SQL_BIT, SQL_BLOB, SQL_BINARY, SQL_VARBINARY and SQL_LONGVARBINARY) are not touched regardless of the mysql_enable_utf8 attribute state.

However, the following code inserts corrupt binary data:

use DBI qw(:sql_types);

open FILE, "test.bin";
binmode FILE;
read(FILE, my $file, 100000000);
close FILE;

$dbh = DBI->connect($dsn,$user,$password,{mysql_enable_utf8 => 1});
$sth = $dbh->prepare("INSERT INTO pdo_blob (the_blob) VALUES(?)");
$sth->bind_param( 1, $file);            # corruption
#$sth->bind_param(1, $file, SQL_BLOB);  # no corruption
$sth->execute();
($fileNew) = $dbh->selectrow_array("SELECT the_blob FROM pdo_blob ORDER BY id DESC LIMIT 1");
print ($file eq $fileNew ? "Equal" : "Not equal");

This change also make it impossible to use the following one-liner as it requires the data type of all placeholders to be specified:
$dbh->do($SQL,undef,@$placeholderValues);

**Comment by nightlight1** [comment]
I understand the underlying motivation to make this change. You are right that is a bug that needs to be fixed.
I wanted to share with you the impact of this change from a user perspective as I believe that this necessary fix comes with the side effect of corrupting data without the user/developer noticing it.
I hope this will make the maintainer of this module consider to change the module in such a way that backwards compatibility is maintained.

**Comment by shadowcat-mst** [comment]
@pali turning on mysql_enable_utf8 has been the least broken option for handling unicode with DBD::mysql for over a decade. It's been basically standard for e.g. any DBIx::Class app that uses unicode w/mysql for most of that time.

The only thing you can do here that won't cause massive corruption of people's production data is to keep it behaving the old way, even if the old way is stupid. You can add a warning that the old way is stupid and that people should consider migrating to the new way, but the old code *must* stay bugwards compatible or you're going to trash people's production data.

This is the sort of backcompat problem that most sucks to have as a maintainer, and I do very much feel for you, but there really ain't any alternative here. Sorry.

**Comment by nightlight1** [comment]
Dear @pali Are you the maintainer of this module? What do you think about the below proposal together with mentioning in the documentation that a value 1 is not sufficient as it lacks encoding of the input statement+parameters but that it remains there for backward compatibility, and value 2 is preferred for correct UTF-8 handling in combination with the note that specifying the datatype of each placeholder value is mandatory for this setting?

Working of 4.041:
mysql_enable_utf8 not set or 0: nothing (default)
mysql_enable_utf8 set to 1: 
1. "SET NAMES utf8" at connect? Yes;
2. Input statement and bind parameters encoded? No
3. Retrieved columns decoded? Yes

Working of 4.042:
mysql_enable_utf8 not set or 0: nothing (default)
mysql_enable_utf8 set to 1: 
1. "SET NAMES utf8" at connect? Yes;
2. Input statement and bind parameters encoded? Yes
3. Retrieved columns decoded? Yes

Proposal for 4.043:
mysql_enable_utf8 not set or 0: nothing (default)
mysql_enable_utf8 set to 1: 
1. "SET NAMES utf8" at connect? Yes;
2. Input statement and bind parameters enoded? No
3. Retrieved columns decoded? Yes
mysql_enable_utf8 set to 2: 
1. "SET NAMES utf8" at connect? Yes;
2. Input statement and bind parameters enoded? Yes
3. Retrieved columns decoded? Yes

**Comment by Grinnz** [comment]
@nightlight1 unfortunately it is not that simple. In the old behavior, the decision whether input parameters were encoded or not is based on the internal state of the perl string, not a clear user visible reason. This is part of the bug that was fixed. Also, retrieved columns were decoded in different circumstances as well. Regardless, you are on the right track, the backwards compatible fix to this would be to restore the behavior of mysql_enable_utf8 and mysql_enable_utf8mb4, and add a new option or option value for the correct behavior.

**Comment by pali** [comment]
@Grinnz And that is still not enough! Even without `mysql_enable_utf8` DBD::mysql damage input strings. See this example with *old* DBD::mysql:

```
$ perl -MDBI -MData::Dumper -e '$Data::Dumper::Useqq = 1; my $dbh = DBI->connect("dbi:mysql:test", "test", undef); my $str2 = "\301" . "\N{U+10FF}"; my $str = substr($str2, 0, 1); print Dumper $dbh->selectall_arrayref("SELECT ?", {}, $str)'
$VAR1 = [
          [
            "\303\201"
          ]
        ];
```

You put `"\301"` on input and you woud get `"\303\201"` on output.

With new DBD::mysql you get correct `"\301"`.

So you cannot support both old buggy behavior and also new perl-correct. I have no idea what else can be done there. Trying to support old buggy behavior would mean to introduce either new bugs or break *non*-UTF-8 support at all.

**Comment by Grinnz** [comment]
Then it sounds like the bugfix with mysql_enable_utf8 disabled has not caused any problems so far, and that can remain. It is only the behavior with mysql_enable_utf8/mb4 changing that has caused the issues in this thread. Would it be possible to restore the old behavior with those switches, and add a new switch or value for the fixed UTF-8 behavior?

**Comment by nightlight1** [comment]
Every day this current version is distributed it causes database corruption in production environments (every INSERT/UPDATE statement involving BLOB's causes corruption).
In the event it is a challenge to find a solution it would be better for the time being to rollback the current  (better) UTF8 implementation and put these changes back in a "development release" until all undesired side effects are covered.

**Comment by Grinnz** [comment]
It is not possible to avoid this "side effect" or "fix" it to recognize blob inputs, for reasons that were discussed earlier in this thread. The old way only accidentally worked for this use case and cannot be relied upon for a real fix. The only way to make it both correct for new code and bugwards compatible for old code is to put the correct behavior on a separate option.

**Comment by mbeijen** [comment]
Absolutely. And this is what I want to propose at this point and actually
put out there soon.

--
Michiel

Op wo 28 jun. 2017 om 19:06 schreef Dan Book <notifications@github.com>

**Comment by pali** [comment]
That mean... if you want to use DBD::mysql (correctly or in new application) you would need to set tons of options. It is not a good. It does not solve this problem, again it just hides it.

**Comment by pali** [comment]
Basically I see there two options:
1) Reupload DBD::mysql 4.041 to cpan as a new version 4.043 and freeze it. Fork DBD::mysql 4.042 as e.g. DBD::mysql2 and develop only this one. It would mean that old DBD::mysql would not receive any security updates nor any other updates.
2) Instruct users who have broken applications to use only old broken DBD::mysql versions. And update only if applications are written to correctly pass data to DBD::mysql.

**Comment by pali** [comment]
Or:
3. Reupload 4.041 as 4.043 and 4.042 as 5.000. New major version would mean that API/ABI could be changed and users/developers needs to verify all changes.

**Comment by sartak** [comment]
Option 1 would work well for us. It would let our users continue running their systems without corruption issues, and would give us a chance to plan and orchestrate a migration to DBD::mysql2.

Options 2 and 3 would not work as well because end users doing installations would get the latest, incompatible DBD::mysql due to how CPAN clients prefer the latest version.

**Comment by pali** [comment]
1. Having occupied DBD::mysql just because applications use `experimental option which may change in future` and cannot be fixed is stupid. Also it would mean that application would use DBD::mysql forever even because it contain security issues (one is open!).

And applications which depends on internal perl state would stop working once something other changes! Such applications are broken and their modules or perl itself should not be updated because such thing can break them!

I'm for option 3. Increasing major version is probably the best idea here. Also other perl modules did such thing in past...

**Comment by kentfredric** [comment]
Making the number on the front get bigger has no technical impact, especially no technical impact in regards to avoiding this breakage for people who install things from CPAN.

That kind of numeric change only affects humans, and then, will only affect humans who install each and every dependency by hand, which is not any majority. 

```
cpanm Foo
> upgraded DBD::mysql
> system broken
```
Is what tends to happen in reality.

**Comment by markuswernig** [comment]
Changing the API in such a drastic way as 4.0.42 does should really not be done in what isn't even a minor version change. 
If the effort required to "fix" the code relying on the old behaviour is large, there is a serious risk of people "freezing" DBD::mysql in their installations at 4.0.41, which would mean that they will stay on that version forever. This is neither stupid nor bad, but just common sense. If the choice is between risking data corruption and sticking to an outdated library, there is little to encourage one to do the former.

But I have one question: @pali is insisting that the "broken" code should be fixed. From what I've read so far this would mean changing each and every insert/update to use `$sth->bind_param($p_num, $bind_value, $bind_type);`for each and every bind parameter.

Is that really so? Is there no other way?

If so, I don't think anybody is going to go through the pain of rewriting large parts of their database handling code (I've done a rough estimate for the application that's mentioned in [my RT issue for this problem](https://rt.cpan.org/Ticket/Display.html?id=121921): We would have to change code in over 500 places...)

And on the other hand it would also be totally against all reasonable programming principles to hard-code the structure of your database tables into your source code. 

That said - if it was really necessary to change something, as the behaviour was inconsistent and not utterly deterministic: Why not change it to a consistent, deterministic behaviour that keeps the old semantics and is then documented? Maybe documenting the current behaviour (that data is encoded or not during insertion based on the utf8 flag on the perl object, if I understand @Grinnz correctly) would already clear up much of the ambiguity and help people clear up their code. 
A second, "more correct" behaviour could then still be added as a separate flag. (I am aware that this would probably haunt the maintainers for some versions yet to come, and I see that it's not a decision easily taken ... but probably it's the easiest way in the long run (i.e. with the least damage in the overall count). 
Forking or bumping the major version will not solve the problem of existing application code breaking.

**Comment by nightlight1** [comment]
What about the following proposal that changes the do statement from:
$dbh->do(''INSERT INTO Books (ID,Content) VALUES (?,?)',undef,5,$blob);
to:
$dbh->do(''INSERT INTO Books (ID,Content) VALUES (?,?)',undef,5,[$blob,SQL_BLOB]);

If the placeholder variable equals a reference to an array it will take the first item as the value, and the second as the DataType. This syntax allows backwards compatibility, and does not come with the necessity to rewrite existing code (only if one desires).

**Comment by Grinnz** [comment]
@markuswernig the "utf8 flag" is an internal implementation detail, and relying on it is the entire reason the old behavior was broken and will result in bugs when seemingly unrelated modules or perl itself is updated. There is no way to fix it other than to tell DBD::mysql when you are sending a value that should not be encoded.

@nightlight1 you will have to ask DBI about that. It is simply not part of the interface.

**Comment by markuswernig** [comment]
@Grinnz 
If this means we have to hard-code each field's data type into the source code, DBD::mysql will become hardly usable any more (feels like going back to the 90ies). Phasing out an "automated" method, that is in wide-spread use, in favour of a "manual" one, which requires massive code changes, just can't be done like that, without a (presumably lengthy) transition period, even if the old method was messy ...

**Comment by Grinnz** [comment]
@markuswernig I think you are missing part of it. You only need to specify the data type for fields that you don't want to be encoded: binary data such as blobs. All text data is still encoded automatically (and more reliably, now).

**Comment by pali** [comment]
@markuswernig perl scalar which contains string is treated as sequence of Unicode code points. If you have binary data in perl, then you store them as sequence of Unicode code points U+00 .. U+FF. If you want to send Unicode string in perl via byte orientated protcol/socket/filehandle you need to load appreciate layer of encoding Unicode to 8bit. And for MySQL there are basically two options: UTF-8 (which includes whole Unicode) and "binary" which includes only U+00 .. U+FF. So you need to tell DBI resp. DBD::mysql how should be treated your perl scalar. This is how perl is working and if you are writing in perl such code, you must know this fact.

**Comment by pali** [comment]
@markuswernig utf8 flag based on which was decision made does **not** say if scalar contains binary data!!!

**Comment by markuswernig** [comment]
@pali 
I know. But obviously it triggered the decision in DBD::mysql up to 4.0.41 whether or not the data to send should be encoded, right? Which does not seem too far-fetched, actually. If you have binary data, just turn off the utf8 flag and it won't get encoded. Seems so simple ... but of course, there's more to the story.

**Comment by Grinnz** [comment]
@markuswernig the utf8 flag is poorly named; it does not indicate whether you have binary data. It is a flag for how perl internally represents the string, and is irrelevant to end users (unless using a buggy XS module like old versions of DBD::mysql). Reliance on it causes issues I have already gone over in this thread, and allows for issues based on where your strings come from. To put a fine point on it, *Any perl string can have the utf8 flag set regardless of its content*.

**Comment by markuswernig** [comment]
@Grinnz 
Yes, I am totally aware of that.

Can we now focus on how to solve the problem?

**Comment by pali** [comment]
@markuswernig perl itself, any perl module or any perl code could add utf8 flag for arbitrary perl scalar/string (even for those which are for user "binary"). And at any time! If your application or perl module depends on such thing, then your perl code is buggy, unpredictable and it is your problem (which you should fix), not problem on module which you depends on.

What everybody should do is **not** to continue produce buggy code.

**Comment by Grinnz** [comment]
@markuswernig nothing involving the utf8 flag is a solution. There is no way to detect automatically if a string should be encoded. It's that simple. You have the same exact situation with printing to a terminal, or sending a HTTP response. You must always be explicit in encoding data to UTF-8 when sending it to a byte protocol. mysql_enable_utf8 simply makes this default to encode, rather than not encode.

**Comment by markuswernig** [comment]
Neither did I write that code, nor do I write code that relies solely on such stuff as the utf8 flag. Please read my comment again. I was proposing to **document** it, that's all.

Can we now focus on how to solve the problem?

**Comment by pali** [comment]
@markuswernig See my PR https://github.com/perl5-dbi/DBD-mysql/pull/119 which documented this problem.

**Comment by Grinnz** [comment]
@markuswernig "Why not change it to a consistent, deterministic behaviour that keeps the old semantics and is then documented?" this is impossible, there is no consistent behavior with those semantics.

**Comment by nightlight1** [comment]
for backwards compatibility you may open a virtual filehandle on a scalar reference and use the file test operators to get a heuristic guess if a scalar is a binary or not.

**Comment by pali** [comment]
> consistent, deterministic behaviour that keeps the old semantics

This is oxymoron. Not only that it is impossible, it does not make sense at all.

> heuristic guess

That has also nothing with consistency.

@nightlight1: What do you mean by that? Because I have not understood your last post. Binary string is *subset* of normal string. See my https://github.com/perl5-dbi/DBD-mysql/issues/117#issuecomment-311784848 . What you wrote for me does not make sense at all.

**Comment by shadowcat-mst** [comment]
@pali what's wrong with mysql_enable_utf8 => 1 keeping doing what it's doing and introducing a new mysql_use_utf8 => 1 or something option for people to use for new code? Seems less hassle for user than having to change their entire dependency

**Comment by markuswernig** [comment]
Another thought: In order to find out the storage data type of a column, would using something like "DESCRIBE $tablename" to determine any binary datatypes be possible before making the decision whether or not to encode the data?
And yes, I also think that a separate option flag for the new behaviour should be introduced for the transition period.

**Comment by Grinnz** [comment]
@markuswernig apart from the obvious issue of overhead, not all binary data is intended to be inserted directly into columns. It may be used in an update or select condition, or passed into some other function (like HEX()) to become a text string.

**Comment by markuswernig** [comment]
@Grinnz 
Yes, but in both cases, if the destination column is *BLOB, *BINARY etc., it should not be encoded before sending.

**Comment by nightlight1** [comment]
I agree with the argument that users will freeze their deployment to 4.041 if it is necessary to make too many changes to their code to upgrade.
With my "heuristic guess" comment I try to find another way (utf8 flag is not usable) to still be able to decide if a specified parameter is binary and therefore not need to be encoded. Yes, this method is not consistent (as the docs make users well aware of this), but that does not mean it is not usable as a fallback scenario if the DataType is missing.

**Comment by pali** [comment]
@markuswernig without parsing whole SQL statement together with information about all schemas it is not possible to know if question mark for argument applies to column or to input of another function. And such parsing is really hard, plus it does not belong to database client. It is part of server.

**Comment by markuswernig** [comment]
@pali 
OK, agreed. Having to parse the entire statement would really not be feasible on the client.

**Comment by Grinnz** [comment]
@nightlight1 freezing the deployment of DBD::mysql was actually recommended several times in this thread. If you are relying on the behavior of DBD::mysql before 4.042 you actually need to freeze the deployment of perl and all installed modules, and only carefully update them to ensure you don't update to something that changes whether your strings are upgraded.

**Comment by markuswernig** [comment]
@Grinnz 
It was mentioned as a possible risk, not as a solution.
Because in the long run this would mean that people would stop using DBD::mysql altogether, and probably invest their resources rather in porting the application to another language, given that the changes necessary because of the behavioural change in 4.0.42 will be considerable.

**Comment by pali** [comment]
> be able to decide if a specified parameter is binary

@nightlight1 In perl it is not possible, as there is no difference. Please look at my https://github.com/perl5-dbi/DBD-mysql/issues/117#issuecomment-311784848

> Because in the long run this would mean that people would stop using DBD::mysql altogether, and probably invest their resources rather in porting the application to another language

@markuswernig So because some people are lazy to understood how perl is working, you suggest that we should to adapt for those people code and let perl module for people who understand module buggy forever? This is wrong decision.

If you write (in any language, in any framework, on any platform) some buggy code, just because you do not know how it is working, then it is your problem.

**Comment by markuswernig** [comment]
@pali 
Please don't get personal. I'm just trying to find solutions.
> If you write (in any language, in any framework, on any platform) some buggy code, just because you do not know how it is working, then it is your problem.

Generally speaking: yes. But in this case, where the module's documentation and behaviour favoured a certain type of coding (that your now call buggy, but which has worked since the very introduction of the `mysql_enable_utf8` flag a decade ago), the call is not so clear.

And even if this were true, it would not solve the problem of a major API change in a sub-minor version update and all the breakage that will come from it.

**Comment by pali** [comment]
Due to root of this problem, introducing any new option which would try to simulate old behavior does not bring old behavior back as it can be affected by any change in code (specially in C/XS part). The only option is to downgrade to old version. There is no safer way how to achieve compatibility.

Also introducing such new option (for compatibility with buggy behavior) does not mean it would work with any set of modules and with any perl version. There is nobody who is doing to maintain such code (if is going to be written) as it is fragile. And as was already wrote any other 3rd module can totally break it.

Plus it affects also mode when mysql_enable_utf8 is not set at all.

Due to all above problems I see only one solution: freeze DBD::mysql at specific version, freeze perl at specific version, freeze all other modules. This is the only way how to achieve 100% compatibility.

If you are happy with less percentage level, you can "unblock" some other modules, probably perl version, but not DBD::mysql.

What is not possible is to have working strings with only codepoints U+80 .. U+FF without current 4.042 patches. And having broken such strings by default (without enabling such options), just because compatibility with buggy applications is stupid. Why having such penalty for new applications? Just because somebody in past introduced in 3rd apps bugs? This is probably reason why not to use Perl/DBI/MySQL.

**Comment by markuswernig** [comment]
@pali 
>  So because some people are lazy to understood how perl is working, you suggest that we should to adapt for those people code and let perl module for people who understand module buggy forever?

No. I'm asking for a reasonable amount of time for fixing the applications.

Keep the behaviour of `mysql_enable_utf8` unchanged, add a new flag enabling the new behaviour (regardless of the value of `mysql_enable_utf8`). Put that in 4.0.43, together with documenting it and pointing out the new recommended usage.
In 4.0.44 start issuing warnings if only `mysql_enable_utf8` is used.
Decide on a timeline and release when to drop the old behaviour.
Put that information in the warning messages.

**Comment by markuswernig** [comment]
> Due to root of this problem, introducing any new option which would try to simulate old behavior does not bring old behavior back as it can be affected by any change in code (specially in C/XS part).

Well, that is exactly how it has been for over 10 years now...

**Comment by shadowcat-mst** [comment]
@pali again, this has worked for some value of worked for 10 years.

You are damaging people's production data.

Please restore the bugs in the old option and pick a new name for the fixed one.

**Comment by nightlight1** [comment]
Dear @pali, are you willing (for now) to publish 4.041 as 4.043 to halt the ongoing corruption of production data?
Let's discuss on how to implement the desired changes after that.

**Comment by pali** [comment]
@nightlight1 As I wrote 4.041 has broken also `mysql_enable_utf8=0`. Applications which needs correct `mysql_enable_utf8=0` and already use 4.042 would be broken by downgrade. So it is not so easy to do such downgrade.

**Comment by markuswernig** [comment]
@pali 
Do you know of any application that relies on the **new** behaviour? Given the amount of time that the old behaviour was out there, and comparing that to the few months that 4.0.42 is out, I strongly doubt that is any relevant percentage.
If yor follow the transition path I suggested (ie. making the new behaviour available under a new name in 4.0.43), all they would have to do is to change the option name from `mysql_enable_utf8` to the new name. Because if you do know of such an app, you can easily inform them about the change.

**Comment by pali** [comment]
Do not know. How do you exactly want to do transition path?

**Comment by markuswernig** [comment]
@pali 
I tried to outline that in [my earlier comment](#issuecomment-311807067)

**Comment by nightlight1** [comment]
the old behavior only incorrectly in rare occasions. This is why it was accepted as a solution for a period of 10 years. The new behavior corrupts all BLOB data of all INSERT & UPDATE statement.
Maybe you don't realize the impact of the but I can tell you it's huge. All binary data that users submitted since we upgraded this module is now corrupt.
Please, be willing to commit the 4.041 version of the files that were changed in 4.042.

**Comment by pali** [comment]
@markuswernig that would mean removal of `mysql_enable_utf8` completely and applications would not be able to be compatible with bold old and new DBD::mysql. This is a bigger change (do not know if better or worse), but would mean that *very* application (even those which used/hacked DBD::mysql to work correctly) needs to be changed. I think this is more drastic change.

@nightlight1 I still think that introducing perl-antipatterns and new bugs (which based on your words could happen only rare) just because of broken applications is wrong.

Also if there are applications which *depends* on broken behavior and perl bugs, they must have proper tests and locked module versions. If not such ecosystem is broken and I do not see reason why other modules should support them. Such decision would mean to continue produce broken/buggy applications.

And if application has absolutely no tests and depends on experimental features, I needs to say that this is problem of such application.

**Comment by markuswernig** [comment]
@pali 
Oh, come on now!
>  that would mean removal of mysql_enable_utf8 completely and applications would not be able to be compatible with bold old and new DBD::mysql.

Your changes have done that already (i.e. there is no way for an application to support both approaches, as the old one does not exist anymore).
If the behaviour triggered by the mysql_enable_utf8 flag was broken and it can't be fixed without breaking other things (as eg. our data), then there is no other way than switching to an new flag name and phasing the old one out. Really.

**Comment by pali** [comment]
> Your changes have done that already (i.e. there is no way for an application to support both approaches, as the old one does not exist anymore).

It is possible to write application to work correctly with old and new DBD::mysql. And it is documented how, see https://github.com/perl5-dbi/DBD-mysql/pull/119

**Comment by nightlight1** [comment]
Dear @pali, my appeal to you is not to discuss what is right or wrong but to pospone that discussion after rolling back the changes.
It is a fact that the changes by 4.042 do more harm than good, and my appeal to you is to work together in preventing data corruption in many organization that is not easy to identify, and not easy to fix as it is a time consuming task to find out which BLOB is corrrupt, and which one is not.

**Comment by pali** [comment]
> my appeal to you is to work together in preventing data corruption

1. write tests for your usage and verify that correct data are written
2. check code that is not using any undefined behavior or something which is fragile or something which is experimental
3. if 2. is yes, then cover everything by test and do proper version locks or verify every update of ecosystem that is not changing something which on which you relay.

There is no other reliable option how to prevent or verify correctness. It can happen with any module or any function.

And if somebody even do not have tests of its production code & data, then sorry it is his problem.

If code/module A is buggy, then A must be fixed. Not everything around A.

**Comment by mbeijen** [comment]
Hi, Michiel here (maintainer of module, together with @CaptTofu )

We knew DBD::mysql had some unicode-related issues.
Pali contributed lots of fixes for that, which is great.
We put out two development releases on CPAN, asking people to test, putting out blog posts and posting on mailing lists. We received no feedback.

We put out a 'stable' release and then it turned out people started running into issues.

As said, the 'enable_utf8' flag was marked as experimental, but it was like that for ten years and it was the only way to get some stuff done with MySQL and Perl and people were relying on that, so I explicitly removed the 'his is experimental and subject to change' remark a while ago (https://github.com/perl5-dbi/DBD-mysql/issues/117#issuecomment-309095916) because I did not want that behaviour to change because it will cause breakage.

So this happened and that is too bad, and now we need a way forward.

I'll release 4.043 *today* which will basically be the same version as 4.041. This means that if people install "latest from cpan" it will still work to the extent that earlier versions of DBD::mysql used to work.
Of course, if there would be in the mean time people using 4.042 and depending on that, they would need to pin that specific version but we expect this to be a large subset of the population. 

While it might be possible to argue that people were relying on bugs in code for correct behaviour, we think it is not acceptable to fix those bugs and make it other peoples problem, in good perl tradition we want to be backwards compatible and we don't want to cause breakage or data corruption for people simply upgrading DBD::mysql.

In our opinion this is the only path forward. After the release of 4.043 we can then add any of the 100+ commits that we had on top of 4.041 that did not break old behaviour, and make a new release out of that.

I'm very sorry for any inconvenience caused!

**Comment by leejo** [comment]
@pali 

> check code that is not using any undefined behavior or something which is fragile or something which is experimental

Unfortunately the `mysql_enable_utf8` option has long since become the de facto way to "fix" UTF-8 issues when using mySQL, regardless of it being buggy, and is used in thousands (hundreds of thousands?) of production apps. The fact that it was marked as experimental in the documentation has been missed by many users as they were referring to blog posts, results of google searches, stack overflow, colleagues who/that didn't know it was experimental or didn't mention this.

Yes we should have all read the docs, but often that doesn't happen. It's become such common knowledge that even published books mention it but don't say it's experimental. Here's just one [example](https://books.google.ch/books?id=daDAnXPnRkcC&pg=PA253&dq=mysql_enable_utf8&hl=en&sa=X&ved=0ahUKEwiI7vys3eLUAhXHOxoKHVdZBjwQuwUIJzAA#v=onepage&q=mysql_enable_utf8&f=false). There are even [tweets](https://twitter.com/b10m/status/417790067697270784) about it.

As @shadowcat-mst says, it sucks to be in this situation, but you really really really *don't* want to be corrupting user's production data. The only viable path here is to keep the old behaviour + warn and add a *new* option for the fix (don't be surprised if you find someone already has `mysql_enable_utf8` set to something other than `1`, `0`, `undef`, or `''`).

FWIW: 10yr+ user of DBD::mysql here, and i would consider myself a reasonably competent developer, but only now finding out that `mysql_enable_utf8` was experimental.

**Comment by shadowcat-mst** [comment]
@mbeijen Thanks. I called what's required 'bugwards compatibility' for a reason (I am *still* annoyed with myself for making DBIx::Class' search() method call wantarray outside of CDBICompat but it is what it is) and am genuinely sympathetic to this because I've had my own similar troubles.

I think part of the problem is pg_enable_utf8 also exists and wasn't experimental, so people just switched it, found it worked, and never saw the warning (I hadn't realised until this thread either). Not your fault but also not something you can do anything about at this stage.

Like I say, all I can think of is requiring a 2 or a different option for the new more consistent behaviour, and maintaining bugwards compatibility for the old behaviour.

I would also very much support a warning-by-default on first connect() for people using the current option, but only if you can suppress it - requiring e.g. a connect option of 'mysql_allow_legacy_utf8_support' or something to suppress the warning so it's obvious to a new maintainer that there's upgrading to do would seem fine to me - see http://p3rl.org/Catalyst::Controller::BindLex for a similarly obnoxious name that helped achieve the desired result.

@pali I'm sure your new code is better/more consistent in general but telling people that it's their fault that code that worked fine for 10 years suddenly doesn't really isn't fair, and I'd ask you to please try and respond to the situation as a sad fact of maintaining a heavily used library rather than a personal attack on your code; you're getting very aggressive towards people who are understandably upset that they just lost production data because of an unforeseen interaction of their code and your fixes, and it's really not fair to them.

**Comment by pali** [comment]
@shadowcat-mst I'm sorry if you understood my words too aggressive or as attack. What I wanted to show is that old buggy behaviour is perl anti-pattern which perl developer does not expect to happen. Also I stated why it does not make sense or it is hard to provide backward compatibility with code which is working pseudo-randomly.

Anyway, now I looked into Changelog of DBD::SQLite and it has exactly same bug. And it was fixed in version 1.43_04 (2014-07-21).

Therefore if you start upgrading DBI application which use both MySQL & SQLite you would get another problem...

As wrote more times, I do not see better option as freezing version of buggy DBD::mysql. It is not possible to have old behavior working in current code. And argument that it worked 10 years is not relevant here if 70% of code was not modified for 10 years...

And because current code is not compatible with new versions of MariaDB, I expect that somebody start doing another big changes which could lead to other problems related to internal perl state.

So for future I do not see other option as fixing DBD::mysql (like in 4.042). And modules/applications which depends on DBD::mysql and are broken would need to be fixed too.

Or stop whole development of DBD::mysql and freeze DBD::mysql forever. There are no other options.

**Comment by pali** [comment]
If developers do not read documentation, release notes, information on mailing list, blogs, etc... there is no way how to tell them to fix their buggy code.

**Comment by Row** [comment]
@pali perhaps you can throw an exception/warning or similar?🤔

**Comment by CaptTofu** [comment]
I'm going to propose that me, Pali, and Michiel discuss how to solve this and get a release out.

**Comment by nightlight1** [comment]
I disagree with one that says it is necessary to revisit all manual pages of all modules an application depends on each time an application is upgraded. This is unfeasable, and undesired. Who starts reading all source code files of a module and read the manual pages of all its dependencies before starting to use it? Nobody.

Please read and respect the Perl policy that have been carefully put together by the community to maintain the quality of Perl as it is today.

> If end-user developers have adopted a bug as a feature, we need to treat it as such.

[http://perldoc.perl.org/perlpolicy.html#BACKWARD-COMPATIBILITY-AND-DEPRECATION](url)

**Comment by pali** [comment]
@nightlight1 you can disagree... but if something is not possible to achieve and fully implement then you would need to take it as is. The only option which I know is freeze of whole module.

**Comment by markuswernig** [comment]
@pali 
>  It is not possible to have old behavior working in current code. 

Which is why the new ("current") code needs to be introduced in a different way than it has been done for 4.0.42. For which it first needs to be backed out and then probably reintroduced in a different manner.

**Comment by markuswernig** [comment]
@pali 
> I'm sorry if you understood my words too aggressive or as attack.

Well, @shadowcat-mst was [not the only one](#issuecomment-311803785) to get that impression ...

**Comment by nightlight1** [comment]
I also read the wording, but I also noticed the many exclamation marks that were added to the manual page of this module. It convinced me that if a person uses these in neutral sentences it may just be his/her common way of expressing and not necessarily meant to be offensive.
For example: people from the Mediterranean appear more enthusiastic in their language than people from Northern Europe.

**Comment by mbeijen** [comment]
So, @CaptTofu and I just released version https://metacpan.org/source/MICHIELB/DBD-mysql-4.043/Changes 4.043 which is the same as 4.041 but with a higher version number.

In the next period, we'll be going to add back fixes that were added after 4.042 and put that out as a development release. We're explicitly going to fix bugs but NOT wanting to break things. If we're going to make changes to how unicode is handled, and if it's going to (also) require changes to existing code, it will be only after users explicitly opting in. We don't want people upgrading their perls or doing an OS upgrade and finding they get data corruption.

The one good thing is there are now lots of eyeballs here. Of course we're going to be extra careful now, but some of you testing upcoming development releases would be more than welcome.

**Comment by Grinnz** [comment]
Thank you. I hope that a new version/option of the corrected unicode handling will be released quickly however. Until then I will not be able to upgrade our DBD::mysql from 4.042 as I do not want to go back to the broken behavior.

**Comment by mbeijen** [comment]
Of course, if you want you're free to pin to 4.042 if that works for you.

Op do 29 jun. 2017 om 23:23 schreef Dan Book <notifications@github.com>

**Comment by markuswernig** [comment]
@mbeijen , @CaptTofu 
Thank you very much! (And of course: Thanks to you and all the other devs for all the work you have already put into DBD::mysql. 
**U rock!**

From what I understand, existing end-user code will now work again.

> Of course we're going to be extra careful now, but some of you testing upcoming development releases would be more than welcome.

I'd be more than happy to help with testing. Where's the best point to be looking (list/blog to be tracking etc.) for changes related to this issue?

**Comment by mbeijen** [comment]
@markuswernig typically we post at dbi-announce (http://lists.perl.org/list/dbi-announce.html) which I'll do today.

**Comment by pali** [comment]
Does it mean that DBD::mysql is now freezed forever? As i have no idea how you want to do future development & still provide that buggy backward compatibility.

**Comment by pali** [comment]
Btw, now look what happened and what you did:
* DBD::mysql has non-working Unicode support forever
* DBD::mysql is incompatible with DBD::Pg and DBD::SQLite which fixed Unicode support in past (in same way as DBD::mysql 4.042)
* 15+ bug fixes were reverted, including fix for amavis
* development of DBD::mysql is going to stagnate, probably full freeze as above nonsense buggy behavior is not possible to achieve with bigger code changes
* nobody would start using DBD::mysql anymore in 2017+ as Unicode support is required

Ask question yourself: Based on above facts, who can be interested in use of DBD::mysql which does not work as expected in perl? New developers? Sure not. Active developers? Not sure, why anymore?

Now I asked for two CVEs for DBD::mysql. We would see who would start fixing security issues based on above fact that buggy behavior cannot be fixed.

And basically everything just because two, tree, four, five (?) people are usable to report bugs to broken application and just started to want have old, broken, non-working and unmaintainable behavior of mysql_enable_utf8 in DBD::mysql driver.

You now **killed** DBD::mysql.

Like @Grinnz said, I already know more people who would need to pin DBD::mysql to 4.042, just because they started to use utf8 and unicode strings correctly, like in whole perl word.

all: Instead of asking for help how you fix your or someone's else buggy application (which I or any other people could help), you rather forced maintainers to go with DBD::mysql to hell.

**Comment by kentfredric** [comment]
> And basically everything just because two, tree, four, five (?) people are usable to report bugs to broken application and just started to want have old, broken, non-working and unmaintainable behavior of mysql_enable_utf8 in DBD::mysql driver.

You're looking at it wrong.

This broke so much code that representatives from 3 different companies had to tell you you'd broken their code.

You have to always see your feedback as the tip of an iceberg, because it is.

And there are probably many more companies out there currently corrupting their data, and won't know about it because they're not checking yet.

Only the very top echelon of developers involve themselves in opensource enough that they *actually communicate* with opensource. The vast majority are *silent consumers*

When you're dealing with software that's purpose in life is to not corrupt data, and have data there tomorrow, you go out of your way not to break that promise.

There's no point in being involved in this sector if you don't care to deliver on that promise.

**Comment by pali** [comment]
Companies which dealing with such data are running on verified infrastructure and every upgrade of core system is tested and reviewed. And in most common scenario they are running some enterprise distribution (like RHEL or their clones) where are versions already freezed.

Nobody in companies is updating random software from random internet places directly on production without any testing.

**Comment by kentfredric** [comment]
> Companies which dealing with such data are running on verified infrastructure and every upgrade of core system is tested and reviewed

No, only the best, most careful, and most expensive of such companies do.

You'd be disturbed how common bad practices are, and practices that exist purely out of lack of time or funding.

Heaven help anyone who dares to make a small business and not really understand their technology.

> Nobody in companies

There's a saying, and its truer than you want to believe: Everyone has a testing rig. 
But only some people also have a production server.

**Comment by Row** [comment]
@kentfredric if you silently introduce the changes of course data will be corrupted! When introducing breaking changes you should deprecate and make sure that old code do not even run. Then developers can choose to stay on an old version or upgrade.

**Comment by leejo** [comment]
> @kentfredric if you silently introduce the changes of course data will be corrupted! When introducing breaking changes you should deprecate and make sure that old code do not even run. Then developers can choose to stay on an old version or upgrade.

@Row - that's exactly what we're all saying in this thread and it's what the very first sentence of the issue report says:

> The utf8 encoding changes have resulted in a regression that has unexpected side effects.

A regression with unexpected side effects is by definition a silent introduction of changes, and we're saying the approach taken (that has since been reverted) needs to be rethought.

**Comment by kentfredric** [comment]
Indeed, its not so much about "no change", its about creating changes such that they avoid breaking code.

And even this is not a "death to DBD::mysql", because there are multiple ways of getting the desired result from different syntax.

There's lots of ways to avoid breaking backwards compatibility.

The most unfortunate thing here is this step wasn't made *before* 4.042 was released.

"Opt in to change[1]" is much more reliable than "make sure upstream didn't break your code every release".

The latter makes sense only if you assume upstream don't care about your data, which is a bad default assumption for a Database Module.

1:  Keeping in mind, "upgrading" and "opting in" are very different things in practice.

**Comment by jacquesg** [comment]
Not intending to hijack this issue, but I've run into **exactly** the same type of problem recently:

dankogai/p5-encode#105

In this case something in core's behaviour silently changed.

**Comment by ap** [comment]
> Based on above facts, who can be interested in use of DBD::mysql which does not work as expected in perl?

You keep saying that, and it keeps not being true.

I ran MySQL in production with “broken” DBD::mysql versions for >10 years, and the content in the database was never wrong. How is that not working as expected?

Upgrading to your “fixes” would have made the content in my database wrong. How is that working as expected?

I was *glad* that the place I worked switched away from MySQL just before your fixes landed in DBD::mysql. You sure did not increase *my* interest in DBD::mysql… sorry to say that.

There is a difference between a bug in the implementation and a bug in the design of an interface. In this particular case, there was a design bug, not an implementation bug. The old interface took an implementation detail (the UTF8 flag in scalars) and made that part of is interface semantics. That’s a bad choice. But it’s an interface design choice. It’s not an implementation bug.

And it can be worked with perfectly reliably if you understand the interface. You and several others wrote repeatedly that the old interface could be broken by upgrades of unrelated modules, and that you had no idea how to write code against such an interface. First of all, that’s only a “could” – there are plenty of cases where that wouldn’t happen. But even then, you absolutely *can* write robust code against an interface that relies on the UTF8 flag. I hope you have heard of `utf8::upgrade` and `utf8::downgrade`? They make it perfectly possible to use a UTF8-flag-reliant interface in a way that is robust against future changes in the dataflow within a program. It’s just onerous.

What you did was to break the contract of a long-standing interface, because its design is bad. The resulting interface design may be better, no dispute – but that’s not a bugfix. It’s interface breakage.

You want a better interface? (Yes, it’s a better interface.) Add it. Don’t break one that already has tons of users.

**Comment by Grinnz** [comment]
@pali I know this situation is frustrating. I think I can speak for others too when I say that I greatly appreciate your contributions of fixes to this very broken distribution. I hope that they can all be reapplied now, just without changing the broken behavior of mysql_enable_utf8/mb4 that so many are relying upon. I also think that mysql_enable_utf8/mb4 should be marked as deprecated when the fixed behavior is added.

**Comment by CaptTofu** [comment]
We will work out a fix that both solves the issue long-term and doesn't 
break things for existing users. I really appreciate the work Pali has 
done and take seriously his concerns as I do those in the community who 
need their code to work. It certainly seems that there is a lot of 
interest in this which will help get perhaps people testing it.

**Comment by Grinnz** [comment]
Is it still planned to reapply the reverted fixes? This distribution now has all of the issues again that were previously solved.

**Comment by leejo** [comment]
Discussion about this seems to have moved to the perl.dbi.users mailing list: https://www.nntp.perl.org/group/perl.dbi.users/2017/08/msg37429.html

More importantly: https://www.nntp.perl.org/group/perl.dbi.users/2017/09/msg37443.html
